toolchain/driver/clang_runner.cpp

```diff
@@ -146,7 +146,28 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
   // Note the subprocessing will effectively call `clang -cc1`, which turns into
   // `carbon-busybox clang -cc1`, which results in an equivalent `clang_main`
   // call.
-  auto cc1_main = [](llvm::SmallVectorImpl<const char*>& cc1_args) -> int {
+  //
+  // Also note that we only do `-disable-free` filtering in the in-process
+  // execution here, as subprocesses leaking memory won't impact this process.
+  auto cc1_main = [enable_leaking = enable_leaking_](
+                      llvm::SmallVectorImpl<const char*>& cc1_args) -> int {
+    // Clang doesn't expose any option to disable injecting `-disable-free` into
+    // the CC1 invocation, or any way to append a flag that undoes it. So to
+    // avoid leaks, we need to edit the `cc1_args` here and remove
+    // `-disable-free`.
+    //
+    // TODO: We should see if upstream would be open to some configuration hook
+    // to suppress this when it is generating the CC1 flags and use that.
+    if (!enable_leaking) {
+      llvm::StringRef flag = "-disable-free";
+      if (const char** it = llvm::find(cc1_args, flag); it != cc1_args.end()) {
+        cc1_args.erase(it);
+      }
+      // There should only ever be one of these flags, but let's be sure.
+      CARBON_CHECK(llvm::find(cc1_args, flag) == cc1_args.end(),
+                   "Found duplicate `-disable-free` flag!");
```

Comment:
Would it make sense to test that the runner doesn't crash with something like `runner.Run({"-c", file.string(), "-o", output.string(), "-disable-free", "-disable-free"});` (sorry, probably not phrasing the flags correctly to get them forwarded)? If we're forwarding flags from the command line, we may need to be resilient to this kind of input.

That also makes me ask, had you considered the slightly different:

```suggestion
      // There should only be one flag instance, but this will handle multiple if present.
      llvm::erase(cc1_args, "-disable-free");
```

Reply:
I hadn't found `llvm::erase`, happy to use that.

I didn't really want to spend the complexity to support removing multiple of these, but I guess that function makes it easy enough.